core/src/main/java/io/grpc/ServiceDescriptor.java

```diff
@@ -32,74 +32,98 @@
 package io.grpc;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 
-import com.google.common.base.Preconditions;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 import javax.annotation.Nullable;
 
 /**
  * Descriptor for a service.
+ *
+ * @since 1.0.0
  */
 public final class ServiceDescriptor {
 
   private final String name;
   private final Collection<MethodDescriptor<?, ?>> methods;
-  private final Object marshallerDescriptor;
+  private final Object schemaDescriptor;
 
+  /**
+   * Constructs a new Service Descriptor.  Users are encouraged to use {@link #newBuilder()}
+   * instead.
+   *
+   * @param name The name of the service
+   * @param methods The methods that are part of the service
+   * @since 1.0.0
+   */
   public ServiceDescriptor(String name, MethodDescriptor<?, ?>... methods) {
-    this(name, null, Arrays.asList(methods));
+    this(name, Arrays.asList(methods));
   }
 
+  /**
+   * Constructs a new Service Descriptor.  Users are encouraged to use {@link #newBuilder()}
+   * instead.
+   *
+   * @param name The name of the service
+   * @param methods The methods that are part of the service
+   * @since 1.0.0
+   */
   public ServiceDescriptor(String name, Collection<MethodDescriptor<?, ?>> methods) {
-    this(name, null, methods);
-  }
-
-  public ServiceDescriptor(String name, Object marshallerDescriptor,
-                           MethodDescriptor<?, ?>... methods) {
-    this(name, marshallerDescriptor, Arrays.asList(methods));
+    this(newBuilder().setName(name).addAllMethods(checkNotNull(methods, "methods")));
   }
 
-  /** Creates a new ServiceDescriptor. */
-  public ServiceDescriptor(String name, Object marshallerDescriptor,
-                           Collection<MethodDescriptor<?, ?>> methods) {
-    this.name = Preconditions.checkNotNull(name, "name");
-    Preconditions.checkNotNull(methods, "methods");
-    validateMethodNames(name, methods);
-    this.marshallerDescriptor = marshallerDescriptor;
-    this.methods = Collections.unmodifiableList(new ArrayList<MethodDescriptor<?, ?>>(methods));
+  private ServiceDescriptor(Builder b) {
+    this.name = checkNotNull(b.name, "name");
```

Comment:
I think this needs to be a IllegalStateException instead of IllegalArgumentException. So checkState.